### PR TITLE
CVSB-3942

### PR DESCRIPTION
**Pull request by vladddragomir**
Ticket that implements TIR test types.

### CODE REVIEW

File: src/providers/test-result/test-result.service.spec.ts [hunk 1]
```diff
@@ -113,13 +113,13 @@ describe('Provider: TestResultService', () => {
     expect(httpService.postTestResult).toHaveBeenCalled();
   });
 
-  it('should return the correct format of the certificateNumber depending on whether it is lec or not', () => {
+  it('should return the correct format of the certificateNumber depending on whether it is tir or not', () => {
```
Comment (KHodderDVSA) [inline src/providers/test-result/test-result.service.spec.ts]: This looks like you removed the LEC  test. Any reason why?

File: src/app/app.enums.ts [hunk 2]
```diff
@@ -473,16 +473,24 @@ export enum LEC_CERTIFICATE_NUMBER_PREFIXES {
...
 
+export enum TIR_CERTIFICATE_NUMBER_PREFIXES {
+  GB_V = 'GB/V',
+  GB_T = 'GB/T'
+}
+
 export enum SPEC_VALUES {
   CERTIFICATE_NUMBER = '67868977',
-  LEC_CERTIFICATE_NUMBER = 'LP67868977',
+  TIR_CERTIFICATE_NUMBER = 'GB/V67868977',
```
Comment (KHodderDVSA) [inline src/app/app.enums.ts]: Connected to comment below - why did you remove the LEC test?

File: src/pages/testing/test-creation/test-create/test-create.spec.ts [hunk 3]
```diff
@@ -221,6 +221,16 @@ describe('Component: TestCreatePage', () => {
     expect(component.getTestTypeStatus(vehicle, testType)).toEqual('Edit');
   });
 
+  it('should have "Edit" status if a TIR test has the certificateNumber exactly 5 digits long', () => {
+    let testType: TestTypeModel = TestTypeDataModelMock.TestTypeData;
```
Comment (KHodderDVSA) [inline src/pages/testing/test-creation/test-create/test-create.spec.ts]: careful with these. mutation on a global variable. better to use clone (lodash, by preference, or ...  if you know it's a flat object)

File: src/pages/testing/test-creation/complete-test/complete-test.ts [hunk 4]
```diff
@@ -55,7 +33,7 @@ export class CompleteTestPage implements OnInit {
   defectsCategories: DefectCategoryReferenceDataModel[];
   isCertificateNumberFocused: boolean;
   today: string;
-  patterns;
+  patterns: typeof REG_EX_PATTERNS = REG_EX_PATTERNS;
```
Comment (KHodderDVSA) [inline src/pages/testing/test-creation/complete-test/complete-test.ts]: ? Don't quite understand what's happening here

File: src/pages/testing/test-creation/complete-test/complete-test.ts [hunk 5]
```diff
@@ -100,7 +81,8 @@ export class CompleteTestPage implements OnInit {
       }
     );
     if (this.vehicleTest.numberOfSeatbeltsFitted && this.testTypeDetails.category === 'B') this.errorIncomplete = false;
-    if (AdrTestTypesData.AdrTestTypesDataIds.indexOf(this.vehicleTest.testTypeId) !== -1 && this.vehicleTest.certificateNumber && this.vehicleTest.certificateNumber.length && this.vehicleTest.certificateNumber.length < 6 && this.errorIncomplete) this.errorIncompleteCertificateNumber = true;
+    if ((AdrTestTypesData.AdrTestTypesDataIds.indexOf(this.vehicleTest.testTypeId) !== -1 && this.vehicleTest.certificateNumber && this.vehicleTest.certificateNumber.length && this.vehicleTest.certificateNumber.length < 6 && this.errorIncomplete) ||
```
Comment (KHodderDVSA) [inline src/pages/testing/test-creation/complete-test/complete-test.ts]: very complex -  split into named functions?

File: src/pages/testing/test-creation/complete-test/complete-test.ts [hunk 6]
```diff
@@ -429,6 +422,10 @@ export class CompleteTestPage implements OnInit {
 
   certificateNumberInputChange(value) {
     this.cdRef.detectChanges();
-    this.vehicleTest.certificateNumber = value.length > 6 ? value.substring(0, 6) : value;
+    if (this.tirTestTypesIds.indexOf(this.vehicleTest.testTypeId) !== -1) {
```
Comment (KHodderDVSA) [inline src/pages/testing/test-creation/complete-test/complete-test.ts]: Should definitely extract this to a common util function. It (and equivalents) get used a lot

File: src/pages/testing/test-creation/test-create/test-create.ts [hunk 7]
```diff
@@ -149,14 +130,21 @@ export class TestCreatePage implements OnInit {
         testType.completionStatus = TEST_COMPLETION_STATUS.EDIT;
         for (let section of testTypeFieldMetadata.sections) {
           for (let input of section.inputs) {
-            if ((!testType[input.testTypePropertyName] && testType[input.testTypePropertyName] !== false) || (AdrTestTypesData.AdrTestTypesDataIds.indexOf(testType.testTypeId) !== -1 && input.testTypePropertyName === TEST_TYPE_INPUTS.CERTIFICATE_NUMBER)) {
+            if ((!testType[input.testTypePropertyName] && testType[input.testTypePropertyName] !== false) ||
+              (AdrTestTypesData.AdrTestTypesDataIds.indexOf(testType.testTypeId) !== -1 && input.testTypePropertyName === TEST_TYPE_INPUTS.CERTIFICATE_NUMBER) ||
```
Comment (KHodderDVSA) [inline src/pages/testing/test-creation/test-create/test-create.ts]: again, put  these in util functions - at least the "isADR/TIRTestType" bits

File: src/pages/testing/test-creation/complete-test/complete-test.html [hunk 8]
```diff
@@ -106,24 +106,24 @@ <h3 item-end class="current-value no-data" *ngIf="!completedFields[input.testTyp
...
+                         [(ngModel)]="vehicleTest[input.testTypePropertyName]"
+                         (ngModelChange)="certificateNumberInputChange($event)"></ion-input>
             </ion-col>
           </ion-row>
         </ion-grid>
-        <p padding-left class="certificate-number-guidance"
-           *ngIf="input.type === testTypeFields.CERTIFICATE_NUMBER_CUSTOM && canDisplaySection(section)">After
-          conducting the test, complete the Low Emissions Certificate and enter the certificate
-          number.</p>
+        <p padding-left class="error-incomplete-certificate-number"
```
Comment (KHodderDVSA) [inline src/pages/testing/test-creation/complete-test/complete-test.html]: presumably this replacement is premised on the LEC certs being generated by us. As I understand it, that's still  not approved - will that block this ticket going through?

File: src/app/app.enums.ts (hunk 2 above)
Comment (KHodderDVSA) [reply src/app/app.enums.ts]: just worked out that this is presumably connected to the removal of the requirement to enter cert details for  LEC certs. Also pending approval - will that prevent this from going in?

File: src/pages/testing/test-creation/test-create/test-create.spec.ts (hunk 3 above)
Comment (vladddragomir) [reply src/pages/testing/test-creation/test-create/test-create.spec.ts]: yes, i should probably use the spread operator {...mock} which will return a copy

File: src/pages/testing/test-creation/complete-test/complete-test.ts (hunk 4 above)
Comment (vladddragomir) [reply src/pages/testing/test-creation/complete-test/complete-test.ts]: patterns is an enum used in the template for the keyboard types that should be displayed for different fields. it didn't have assigned the specific enum to it so it was not recognised in the template.

File: src/pages/testing/test-creation/complete-test/complete-test.ts (hunk 5 above)
Comment (vladddragomir) [reply src/pages/testing/test-creation/complete-test/complete-test.ts]: ok

File: src/pages/testing/test-creation/complete-test/complete-test.ts (hunk 6 above)
Comment (vladddragomir) [reply src/pages/testing/test-creation/complete-test/complete-test.ts]: ok

File: src/pages/testing/test-creation/test-create/test-create.ts (hunk 7 above)
Comment (vladddragomir) [reply src/pages/testing/test-creation/test-create/test-create.ts]: ok

File: src/pages/testing/test-creation/complete-test/complete-test.html (hunk 8 above)
Comment (vladddragomir) [reply src/pages/testing/test-creation/complete-test/complete-test.html]: this custom field should have also been removed due to the lec test types iteration, but i missed it by God's will :))) because i would have need it again now

File: src/providers/test-result/test-result.service.spec.ts (hunk 1 above)
Comment (vladddragomir) [reply src/providers/test-result/test-result.service.spec.ts]: same God's will

File: src/app/app.enums.ts (hunk 2 above)
Comment (vladddragomir) [reply src/app/app.enums.ts]: same God's will

File: src/pages/testing/test-creation/test-create/test-create.spec.ts (hunk 3 above)
Comment (vladddragomir) [reply src/pages/testing/test-creation/test-create/test-create.spec.ts]: done

File: src/pages/testing/test-creation/complete-test/complete-test.ts (hunk 5 above)
Comment (vladddragomir) [reply src/pages/testing/test-creation/complete-test/complete-test.ts]: done

File: src/pages/testing/test-creation/complete-test/complete-test.ts (hunk 6 above)
Comment (vladddragomir) [reply src/pages/testing/test-creation/complete-test/complete-test.ts]: done

File: src/pages/testing/test-creation/test-create/test-create.ts (hunk 7 above)
Comment (vladddragomir) [reply src/pages/testing/test-creation/test-create/test-create.ts]: done
